handwriting_line_recognition.py

```diff
@@ -0,0 +1,392 @@
+            i += 1
+    return image, label_encoded
+
+def augment_transform(image, label):
+    '''
+    This function randomly:
+        - translates the input image by +-width_range and +-height_range (percentage).
+        - scales the image by y_scaling and x_scaling (percentage)
+        - shears the image by shearing_factor (radians)
+    '''
+
+    ty = random.uniform(-random_y_translation, random_y_translation)
+    tx = random.uniform(-random_x_translation, random_x_translation)
+
+    sx = random.uniform(1. - random_y_scaling, 1. + random_y_scaling)
+    sy = random.uniform(1. - random_x_scaling, 1. + random_x_scaling)
+
+    s = random.uniform(-random_shearing, random_shearing)
+
+    st = skimage_tf.AffineTransform(scale=(sx, sy),
+                                    shear=s,
+                                    translation=(tx*image.shape[1], ty*image.shape[0]))
+    augmented_image = skimage_tf.warp(image, st, cval=1.0)
+    return transform(augmented_image*255., label)
+
+def decode(prediction):
+    '''
+    Returns the string given one-hot encoded vectors.
+    '''
+    results = []
+    for word in prediction:
+        result = []
+        for i, index in enumerate(word):
+            if i < len(word) - 1 and word[i] == word[i+1] and word[-1] != -1: #Hack to decode label as well
+                continue
+            if index == len(alphabet_dict) or index == -1:
+                continue
+            else:
+                result.append(alphabet_encoding[int(index)])
+        results.append(result)
+    words = [''.join(word) for word in results]
+    return words
+
+def run_epoch(e, network, dataloader, trainer, log_dir, print_name, update_network, save_network):
+    '''
+    Run one epoch to train or test the CNN-biLSTM network
+    
+    Parameters
+    ----------
+        
+    e: int
+        The epoch number
+
+    network: nn.Gluon.HybridSequential
+        The CNN-biLSTM network
+
+    dataloader: gluon.data.DataLoader
+        The train or testing dataloader that is wrapped around the iam_dataset
+    
+    log_dir: Str
+        The directory to store the log files for mxboard
+
+    print_name: Str
+        Name to print for associating with the data. usually this will be "train" and "test"
+    
+    update_network: bool
+        Boolean to indicate whether or not the network should be updated. Update_network should only be set to true for the training data
+
+    save_network: bool
+        Boolean to indicate whether or not to save the network. 
+
+    Returns
+    -------
+    
+    epoch_loss: float
+        The loss of the current epoch
+    '''
+
+    total_loss = nd.zeros(1, ctx)
+    for i, (x, y) in enumerate(dataloader):
+        x = x.as_in_context(ctx)
+        y = y.as_in_context(ctx)
+
+        with autograd.record():
+            output = network(x)
+            loss_ctc = ctc_loss(output, y)
+
+        if update_network:
+            loss_ctc.backward()
+            trainer.step(x.shape[0])
+
```

Comment:
looks like there is no blocking call in your training loop, could be a cause of having the crashes you were talking about

Reply:
is total_loss.asscalar() a blocking call?

Reply:
it is blocking indeed

Reply:
so nd.waitall() after each epoch?